### ISSUE TITLE
Tweens cause memory leak

### ISSUE THREAD

**Issue by markdibarry** [godotengine/godot#61340]
### Godot version

4.0.dev (a80793f)

### System information

Windows 10

### Issue description

This has only been tested in C#, but may exist in GDScript as well. Similar to #57731 and #61341, when calling `Node.CreateTween()`, `SceneTree.CreateTween()` or `Tween.TweenProperty()`, objects are created and never freed, and must be disposed manually.

### Steps to reproduce

Call any of the methods mentioned above.

**Comment by KoBeWi** [comment]
Duplicate of #52699

**Comment by KoBeWi** [comment]
On a second thought, this might be another issue, as the other one seems specific to GDScript.

**Comment by markdibarry** [comment]
Just an update that this is in 3.5 as well with `SceneTreeTween`.

**Comment by raulsntos** [comment]
I'll quote a response from **neikeq** in a related issue: https://github.com/godotengine/godot/issues/53172#issuecomment-929336671

> The garbage collector doesn't necessarily collect resources right away. There were similar issues to this in the past, and the conclusion was that the GC simply didn't collect them yet.

Try executing the following code after creating the tween:

```csharp
GC.Collect(GC.MaxGeneration);
GC.WaitForPendingFinalizers();
```

If that frees the memory it means there is no leak, the GC just hasn't collected the garbage yet (so just wait or call `Dispose` manually). If after executing the code the memory is not freed then there might be a memory leak.

**Comment by markdibarry** [comment]
@raulsntos I could have sworn it was already mentioned on some of these, but if not, I'm sorry. So there's no uncertainty: Similar to #61341, #57731 and (the much bigger) #61911, these require manual disposal. The code provided above will not free the memory and the GC will never collect them.

**Comment by raulsntos** [comment]
I don't know about the other mentioned issues, those may or may not be a memory leak, the code provided should be used to check if they are.

> The code provided above will not free the memory and the GC will never collect them.

Can I ask how you are checking the memory usage? (using Godot's monitors, your IDE...).
Also, have you checked if the code frees the memory in both 4.0 and 3.x or only 4.0? And if you have a MRP it'd be great if you could upload it.

**Comment by markdibarry** [comment]
> I don't know about the other mentioned issues
Just mentioning here, at least, that I used the same code to force collection to verify all the memory leaks I've reported.

I don't believe there's an IDE that can do profiling with Godot atm (though I'd be happy to be wrong!), so I'm just using the Object count in the Godot monitors. I've verified that the memory does not free in both 4.0 with `Tween` and `PropertyTweener` and same with 3.5 with `SceneTreeTween` and `PropertyTweener`.

I'll make a MRP. It's a pretty quick setup.

**Comment by markdibarry** [comment]
@raulsntos Here you go. This is for 4.0:
[TweenBug4.zip](https://github.com/godotengine/godot/files/8934136/TweenBug4.zip)

**Comment by raulsntos** [comment]
I've been able to reproduce this in 4.0 but not in 3.5 RC 4.

I'd say all the mentioned issues are the same, basically C# not freeing `RefCounted<T>` instances (`Reference<T>` in 3.x). The issues mention `Tween`, `Material` and `Resource` all of which end up inheriting from `RefCounted<T>` so I think that's where the problem is.

The problem with `RefCounted<T>` in C# is that it's difficult to manage the reference count (see https://github.com/godotengine/godot-proposals/issues/3820 for more details). The current implementation may not be ideal but I think it should work and as I said I could not reproduce the issue in `3.x` so maybe it's just a regression in `master`.

**Comment by markdibarry** [comment]
Ah, that makes sense why `Tween`, `Material`, and `Resource` all are having the same problems in 4.0. Since we've got to the root of it, it probably makes sense to close my other tickets and make a general `RefCounted<T>` bug ticket. I hope that can be fixed soon!

You're right it's actually not reproducible on 3.5. I was checking it against 4.0 and I accidentally set the force collect invocation to the wrong input mapping. Once I fixed that, the force collection cleaned up the old objects. I'll close this as user error.
